graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/tracing/FederatedTracingInstrumentation.java

```diff
@@ -314,6 +291,138 @@ private Timestamp getNowTimestamp() {
+                // considered bad practice.
+                Lock l = treeLock.writeLock();
+                l.lock();
+                try {
+                    isFinalized = false;
+                } finally {
+                    l.unlock();
+                }
+            }
+
+            /**
+             * Edit builder for the node at the given path (creating it and its parents if needed).
+             */
+            public void editBuilder(ExecutionPath path, Consumer<Reports.Trace.Node.Builder> builderConsumer) {
+                Lock l = treeLock.readLock();
+                l.lock();
+                try {
+                    if (isFinalized) {
+                        throw new RuntimeException("Cannot edit builder after protobuf conversion.");
+                    }
+                    Node node = getOrCreateNode(path);
+                    synchronized (node.builder) {
+                        builderConsumer.accept(node.builder);
+                    }
+                } finally {
+                    l.unlock();
+                }
+            }
+
+            /**
+             * Get node for the given path in nodesByPath (creating it and its parents if needed).
+             */
+            @NotNull
+            private Node getOrCreateNode(ExecutionPath path) {
+                // Fast path for when the node already exists.
+                Node current = nodesByPath.get(path);
+                if (current != null) return current;
+
+                // Find the latest ancestor that exists in the map.
+                List<Object> pathSegments = path.toList();
+                int currentSegmentIndex = pathSegments.size();
+                while (current == null) {
+                    if (currentSegmentIndex <= 0) {
+                        // The root path's node is inserted at construction time, so this shouldn't
+                        // happen.
+                        throw new RuntimeException("root path missing from nodesByPath?");
+                    }
+                    currentSegmentIndex--;
+                    ExecutionPath currentPath = ExecutionPath.fromList(pathSegments.subList(0, currentSegmentIndex));
+                    current = nodesByPath.get(currentPath);
+                }
+
+                // Travel back down to the requested node, creating child nodes along the way as
+                // needed.
+                for (; currentSegmentIndex < pathSegments.size(); currentSegmentIndex++) {
+                    Node parent = current;
+                    ExecutionPath childPath = ExecutionPath.fromList(pathSegments.subList(0, currentSegmentIndex + 1));
+                    Object childSegment = pathSegments.get(currentSegmentIndex);
+
+                    // Note that the correctness of this statement depends on ConcurrentHashMap's
+                    // guarantee that computeIfAbsent will execute the lambda at most once.
```

Comment:
> Does that guarantee also include "if the lambda is called, then its result will actually get in the map"? Looks like it, I think. But that's the real guarantee we care about, right?

Yes, that would appear to be the case from the docs.

The docs for `ConcurrentHashMap.computeIfAbsent()` say that "the entire method invocation is performed atomically, so the function is applied at most once per key." The method itself is described as "if the specified key is not already associated with a value, attempts to compute its value using the given mapping function and enters it into this map unless null."

I would take the "atomically" part to mean that
- checking if the key has a preexisting value
- executing the lambda (if no value was found)
- inserting the resulting value (if that value is non-null)
- returning the new (or preexisting) value
appears to occur as one indivisible step to callers of `ConcurrentHashMap`.

And accordingly, if the lambda is executed and returns a non-null value, it should indeed appear to be inserted after the call returns.

Other method calls on `ConcurrentHashMap` will either see the null (in which case that call happened before the `computeIfAbsent()` call), or they'll see the new value (in which case it happened after), but they won't see any steps in between/be interleaved, because that would break atomicity of `computeIfAbsent()`.

We can make similar reasoning for `ConcurrentHashMap.putIfAbsent()`, which is also said to be atomic. For that method,
- checking if the key has a preexisting value
- inserting the given value (if no value was found)
- returning the preexisting value (or null)
similarly appears to occur as one indivisible step to callers of `ConcurrentHashMap`.

> A simpler option that doesn't rely on this subtler property would be to always create a Node and a Builder, use putIfAbsent instead of computeIfAbsent, and call parent.children.add if putIfAbsent returns null. (It's a shame computeIfAbsent can't tell you if the result was used or not — well, I guess the lambda could remember if it was called, but that goes back to my previous question.) I think that is simpler to reason about, and the extra allocations aren't likely to occur much — only in the case where multiple nodes are being concurrently created with missing common ancestors, and those common ancestors were still not around when doing the while loop above.

I'd be fine with moving to `ConcurrentHashMap.putIfAbsent()`. `computeIfAbsent()` can notably block other callers of `computeIfAbsent()`, and we can avoid that if we use the `putIfAbsent()` approach as you've described.